sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala

```diff
@@ -19,274 +19,575 @@ package org.apache.spark.sql.execution.streaming
+ * specific indexInList from the RocksDB value. (In the future, we could write a custom
+ * merge operator that can handle tombstones for deleted indexes, but RocksDB doesn't
+ * support custom merge operators written in Java/Scala.)
+ *
+ * Instead, we manage expiration per grouping key instead. Our secondary index will look
+ * like (expirationMs, groupingKey) -> EMPTY_ROW. This way, we can quickly find all the
+ * grouping keys that contain at least one element that has expired.
+ *
+ * There is some trickiness here, though. Suppose we have an element key `k` that
+ * has a list with one value `v1` that expires at time `t1`. Our primary index looks like
+ * k -> [v1]; our secondary index looks like [(t1, k) -> EMPTY_ROW]. Now, we add another
+ * value to the list, `v2`, that expires at time `t2`. The primary index updates to be
+ * k -> [v1, v2]. However, how do we update our secondary index? We already have an entry
+ * in our secondary index for `k`, but it's prefixed with `t1`, which we don't know at the
+ * time of inserting `v2`.
+ *
+ * So, do we:
+ *    1. Blindly add (t2, k) -> EMPTY_ROW to the secondary index?
+ *    2. Delete (t1, k) from the secondary index, and then add (t2, k) -> EMPTY_ROW?
+ *
+ * We prefer option 2 because it avoids us from having many entries in the secondary
+ * index for the same grouping key. But when we have (t2, k), how do we know to delete
+ * (t1, k)? How do we know that t1 is prefixing k in the secondary index?
+ *
+ * To solve this, we introduce another index that maps element key to the the
+ * minimum expiry timestamp for that element key. It is called the min-expiry index.
+ * With it, we can know how to update our secondary index, and we can still range-scan
+ * the TTL index to find all the keys that have expired.
+ *
+ * In our previous example, we'd use this min-expiry index to tell us whether `t2` was
+ * smaller than the minimum expiration on-file for `k`. If it were smaller, we'd update
+ * the TTL index, and the min-expiry index. If not, then we'd just update the primary
+ * index. When the batchTimestampMs exceeded `t1`, we'd know to clean up `k` since it would
+ * contain at least one expired value. When iterating through the many values for this `k`,
+ * we'd then be able to find the next minimum value to insert back into the secondary and
+ * min-expiry index.
+ *
+ * All of this logic is implemented by the updatePrimaryAndSecondaryIndices function.
  */
-abstract class SingleKeyTTLStateImpl(
-    stateName: String,
-    store: StateStore,
-    keyExprEnc: ExpressionEncoder[Any],
-    ttlExpirationMs: Long)
-  extends TTLState {
-
-  import org.apache.spark.sql.execution.streaming.StateTTLSchema._
-
-  private val ttlColumnFamilyName = "$ttl_" + stateName
-  private val keySchema = getSingleKeyTTLRowSchema(keyExprEnc.schema)
-  private val keyTTLRowEncoder = new SingleKeyTTLEncoder(keyExprEnc)
+abstract class OneToManyTTLState(
```

Comment:
Yeah likewise I said in above, I'm saying this because we don't know we would solve every one-to-many values case to put and merge, and this class is tied to this. If we move over the responsibility of updating primary index (key-value) it's probably sufficient to generalize, but I guess it's probably non-trivial to decouple things.

I'm not suggesting we need to generalize. I'm OK with this; I'd see more specific name if it's specific, but I'm also OK with just calling out in the class doc.